### PR TITLE
Added semantic text chunker

### LINKED ISSUE #44: Evaluate a new feature to split (i.e., chunk) text based on the semantics of the text
Instead of using a global fixed chunk size for RAG, try splitting based on the semantics of the text.

[@GregKamradt](https://twitter.com/GregKamradt) proposed a super simple method to split long documents based on embedding similarity between sentences, with an auto-tuned threshold.

If there’s sufficiently low similarity between two sentences, then there’s a break and you can split!

This seems to be implemented in LlamaPack 👉🏻 [@llama_index](https://twitter.com/llama_index) abstractions, so you can use this to power advanced RAG pipelines.

Semantic Chunking LlamaPack: https://llamahub.ai/l/llama_packs-node_parser-semantic_chunking?from=all

Notebook: [https://github.com/run-llama/llama-hub/blob/main/llama_hub/llama_packs/node_parser/semantic_chunking/semantic_chunking.ipynb](https://t.co/QxeKncUYZZ)

Let's evaluate if we can add this to ENMARCHA.

### PR DESCRIPTION
**Pull request by LuisM000**
This pull request addresses issue #44 and introduces a new feature for semantic text splitting. The main additions include:

1. **New Interface and Implementation:**
   - Added a new interface `Encamina.Enmarcha.AI.Abstractions.ISemanticTextSplitter` for semantic text splitting.
   - Implemented the interface with `Encamina.Enmarcha.AI.SemanticTextSplitter`, which utilizes semantic analysis to split text into meaningful chunks.

2. **Mathematical Utility Class:**
   - Added a new utility class `Encamina.Enmarcha.Core.MathUtils` containing methods for mathematical operations. This class includes functionalities like percentile calculation, standard deviation, quartiles, interquartile range, and median.

### Purpose
The new semantic text splitter enhances the capabilities of text processing, allowing for more sophisticated analysis and understanding of text structures. The addition of `MathUtils` provides a convenient set of mathematical operations for future use within the project.

### CODE REVIEW

**Review by rliberoff — DISMISSED** [review]

**Review by adiazcan — COMMENTED** [review]
@LuisM000 check sonar warnings, please

**Comment by rliberoff** [comment]
> 

@adiazcan We should check some of those rules in Sonar. Some are outdated, like the maximum number of characters per line.

File: src/Encamina.Enmarcha.AI.Abstractions/ISemanticTextSplitter.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+public interface ISemanticTextSplitter
+{
+    /// <summary>
+    /// Splits the input text based on semantic content.
+    /// </summary>
+    /// <param name="text">The input text to be split.</param>
+    /// <param name="embeddingsGenerator">A function to generate embeddings for a list of strings.</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
+    /// <returns>A collection of text splits.</returns>
+    Task<IEnumerable<string>> SplitAsync(string text, Func<IList<string>, CancellationToken, Task<IList<ReadOnlyMemory<float>>>> embeddingsGenerator, CancellationToken cancellationToken = default);
```
Comment (rliberoff) [inline src/Encamina.Enmarcha.AI.Abstractions/ISemanticTextSplitter.cs]: Just remove the `default` in the `cancellationToken` to solve the Sonar warning.

**Review by rliberoff — CHANGES_REQUESTED** [review]

File: src/Encamina.Enmarcha.AI.Abstractions/ISemanticTextSplitter.cs (hunk 1 above)
Comment (LuisM000) [reply src/Encamina.Enmarcha.AI.Abstractions/ISemanticTextSplitter.cs]: Fixed!

**Comment by LuisM000** [comment]
> @LuisM000 check sonar warnings, please

As we discussed, I have corrected the Sonar errors except for the one related to the maximum number of characters per line (which we will modify in the Sonar rule).

**Comment by LuisM000** [comment]
> > @LuisM000 check sonar warnings, please
> 
> As we discussed, I have corrected the Sonar errors except for the one related to the maximum number of characters per line (which we will modify in the Sonar rule).

@rliberoff has already modified it in Sonar, I have rerun the job, and now we no longer have any warnings :)

**Review by rliberoff — APPROVED** [review]

**Review by dianabisbe — APPROVED** [review]

**Pull request merged**